File: 백준/Platinum/5719. 거의 최단 경로/거의 최단 경로.java
import java.util.*;
import java.io.*;

public class Main {
	public int dijkstra(int[][] graph, int start, int end) {
		int graphLen = graph.length;
		
		int[] minDists = new int[graphLen];
		Arrays.fill(minDists, Integer.MAX_VALUE);
		minDists[start] = 0;
		
		List<List<Integer>> visited = new ArrayList<List<Integer>>();
		for (int i = 0; i < graphLen; i++) {
			visited.add(new ArrayList<Integer>());
		}
		
		PriorityQueue<int[]> heap = new PriorityQueue<int[]>((a, b) -> a[1] - b[1]);
		heap.add(new int[] {start, 0});
		
		while (heap.isEmpty() != true) {
			int[] cur = heap.poll();
			int curNode = cur[0];
			int curDist = cur[1];
			
			if (curDist > minDists[curNode]) {
				continue;
			}
			
			for (int nextNode = 0; nextNode < graphLen; nextNode++) {
				int dist = graph[curNode][nextNode];
				
				if (dist == 0) {
					continue;
				}
				
				int nextDist = dist + curDist;
				
				if (nextDist < minDists[nextNode]) {
					visited.set(nextNode, new ArrayList<Integer>());
					visited.get(nextNode).add(curNode);
					
					minDists[nextNode] = nextDist;
					
					heap.add(new int[] {nextNode, nextDist});
				}
				
				else if (nextDist == minDists[nextNode]) {
					visited.get(nextNode).add(curNode);
				}
			}
		}
		
		Deque<Integer> stk = new LinkedList<Integer>();
		stk.add(end);
		
		while (stk.isEmpty() != true) {
			int e = stk.getLast();
			int visitLen = visited.get(e).size();
			
			if (visitLen == 0) {
				stk.pollLast();
				continue;
			}
			
			for (int i = 0; i < visitLen; i++) {
				int s = visited.get(e).get(i);
				
				if (graph[s][e] != 0) {
					graph[s][e] = 0;
					stk.add(s);
					break;
				}
				
				else if (i == visitLen - 1) {
					stk.pollLast();
				}
			}
		}
		
		return minDists[end];
	}
	
	public void solution() throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
		StringTokenizer st;
		
		int N, M;
		while (true) {
			st = new StringTokenizer(br.readLine());
			N = Integer.parseInt(st.nextToken());
			M = Integer.parseInt(st.nextToken());
			
			if (N == 0) {
				break;
			}
			
			st = new StringTokenizer(br.readLine());
			int S, D;
			S = Integer.parseInt(st.nextToken());
			D = Integer.parseInt(st.nextToken());
			
			int[][] graph = new int[N][N];
			for (int i = 0; i < M; i++) {
				st = new StringTokenizer(br.readLine());
				int U, V, P;
				U = Integer.parseInt(st.nextToken());
				V = Integer.parseInt(st.nextToken());
				P = Integer.parseInt(st.nextToken());
				
				graph[U][V] = P;
			}
			
			dijkstra(graph, S, D);
			int result = dijkstra(graph, S, D);
			if (result == Integer.MAX_VALUE) {
				result = -1;
			}
			
			bw.write(result + "\n");
		}
		
		bw.flush();
		br.close();
		bw.close();
	}
	
	public static void main(String[] args) throws Exception {
		new Main().solution();
	}
}

File: 백준/Platinum/5719. 거의 최단 경로/거의 최단 경로.py
import heapq
import sys
from collections import deque, defaultdict

input = sys.stdin.readline


def dijkstra(graph, start, exc):
    visited = [[] for _ in range(len(graph))]
    min_dists = [float('inf') for _ in range(len(graph))]
    min_dists[start] = 0

    hq = [[0, start]]
    while hq:
        cur_dist, cur_node = heapq.heappop(hq)
        if cur_dist > min_dists[cur_node]:
            continue

        for dist, next_node in graph[cur_node]:
            if next_node in exc[cur_node]:
                continue

            next_dist = cur_dist + dist

            if next_dist < min_dists[next_node]:
                min_dists[next_node] = next_dist
                visited[next_node] = [cur_node]
                heapq.heappush(hq, [next_dist, next_node])

            elif next_dist == min_dists[next_node]:
                visited[next_node].append(cur_node)

    return visited, min_dists


while True:
    N, M = map(int, input().split())

    if not N and not M:
        break

    S, D = map(int, input().split())

    graph = [[] for _ in range(N)]
    for _ in range(M):
        U, V, P = map(int, input().split())
        graph[U].append([P, V])

    exc = defaultdict(set)
    min_visited, min_dists = dijkstra(graph, S, exc)

    dq = deque([D])
    while dq:
        next_node = dq.popleft()

        for cur_node in min_visited[next_node]:
            if next_node in exc[cur_node]:
                continue

            exc[cur_node].add(next_node)
            dq.append(cur_node)

    sec_visited, sec_min_dists = dijkstra(graph, S, exc)
    sec_min_dist = sec_min_dists[D]
    print(sec_min_dist if sec_min_dist != float('inf') else -1)